// @workos-inc/node/lib/feature-flags/feature-flags.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jest_fetch_mock_1 = __importDefault(require("jest-fetch-mock"));
const test_utils_1 = require("../common/utils/test-utils");
const workos_1 = require("../workos");
const list_feature_flags_json_1 = __importDefault(require("./fixtures/list-feature-flags.json"));
const get_feature_flag_json_1 = __importDefault(require("./fixtures/get-feature-flag.json"));
const enable_feature_flag_json_1 = __importDefault(require("./fixtures/enable-feature-flag.json"));
const disable_feature_flag_json_1 = __importDefault(require("./fixtures/disable-feature-flag.json"));
const workos = new workos_1.WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');
describe('FeatureFlags', () => {
    beforeEach(() => jest_fetch_mock_1.default.resetMocks());
    describe('listFeatureFlags', () => {
        describe('without any options', () => {
            it('returns feature flags and metadata', () => __awaiter(void 0, void 0, void 0, function* () {
                (0, test_utils_1.fetchOnce)(list_feature_flags_json_1.default);
                const { data, listMetadata } = yield workos.featureFlags.listFeatureFlags();
                expect((0, test_utils_1.fetchSearchParams)()).toEqual({
                    order: 'desc',
                });
                expect((0, test_utils_1.fetchURL)()).toContain('/feature-flags');
                expect(data).toHaveLength(3);
                expect(data[0]).toEqual({
                    object: 'feature_flag',
                    id: 'flag_01EHQMYV6MBK39QC5PZXHY59C5',
                    name: 'Advanced Dashboard',
                    slug: 'advanced-dashboard',
                    description: 'Enable advanced dashboard features',
                    tags: ['ui'],
                    enabled: true,
                    defaultValue: false,
                    createdAt: '2024-01-01T00:00:00.000Z',
                    updatedAt: '2024-01-01T00:00:00.000Z',
                });
                expect(listMetadata).toEqual({
                    before: null,
                    after: 'flag_01EHQMYV6MBK39QC5PZXHY59C7',
                });
            }));
        });
        describe('with the before option', () => {
            it('forms the proper request to the API', () => __awaiter(void 0, void 0, void 0, function* () {
                (0, test_utils_1.fetchOnce)(list_feature_flags_json_1.default);
                const { data } = yield workos.featureFlags.listFeatureFlags({
                    before: 'flag_before_id',
                });
                expect((0, test_utils_1.fetchSearchParams)()).toEqual({
                    before: 'flag_before_id',
                    order: 'desc',
                });
                expect((0, test_utils_1.fetchURL)()).toContain('/feature-flags');
                expect(data).toHaveLength(3);
            }));
        });
        describe('with the after option', () => {
            it('forms the proper request to the API', () => __awaiter(void 0, void 0, void 0, function* () {
                (0, test_utils_1.fetchOnce)(list_feature_flags_json_1.default);
                const { data } = yield workos.featureFlags.listFeatureFlags({
                    after: 'flag_after_id',
                });
                expect((0, test_utils_1.fetchSearchParams)()).toEqual({
                    after: 'flag_after_id',
                    order: 'desc',
                });
                expect((0, test_utils_1.fetchURL)()).toContain('/feature-flags');
                expect(data).toHaveLength(3);
            }));
        });
        describe('with the limit option', () => {
            it('forms the proper request to the API', () => __awaiter(void 0, void 0, void 0, function* () {
                (0, test_utils_1.fetchOnce)(list_feature_flags_json_1.default);
                const { data } = yield workos.featureFlags.listFeatureFlags({
                    limit: 10,
                });
                expect((0, test_utils_1.fetchSearchParams)()).toEqual({
                    limit: '10',
                    order: 'desc',
                });
                expect((0, test_utils_1.fetchURL)()).toContain('/feature-flags');
                expect(data).toHaveLength(3);
            }));
        });
    });
    describe('getFeatureFlag', () => {
        it('requests a feature flag by slug', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)(get_feature_flag_json_1.default);
            const subject = yield workos.featureFlags.getFeatureFlag('advanced-dashboard');
            expect((0, test_utils_1.fetchURL)()).toContain('/feature-flags/advanced-dashboard');
            expect(subject).toEqual({
                object: 'feature_flag',
                id: 'flag_01EHQMYV6MBK39QC5PZXHY59C5',
                name: 'Advanced Dashboard',
                slug: 'advanced-dashboard',
                description: 'Enable advanced dashboard features',
                tags: ['ui'],
                enabled: true,
                defaultValue: false,
                createdAt: '2024-01-01T00:00:00.000Z',
                updatedAt: '2024-01-01T00:00:00.000Z',
            });
        }));
    });
    describe('enableFeatureFlag', () => {
        it('enables a feature flag by slug', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)(enable_feature_flag_json_1.default);
            const subject = yield workos.featureFlags.enableFeatureFlag('advanced-dashboard');
            expect((0, test_utils_1.fetchURL)()).toContain('/feature-flags/advanced-dashboard/enable');
            expect((0, test_utils_1.fetchMethod)()).toBe('PUT');
            expect(subject.enabled).toBe(true);
        }));
    });
    describe('disableFeatureFlag', () => {
        it('disables a feature flag by slug', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)(disable_feature_flag_json_1.default);
            const subject = yield workos.featureFlags.disableFeatureFlag('advanced-dashboard');
            expect((0, test_utils_1.fetchURL)()).toContain('/feature-flags/advanced-dashboard/disable');
            expect((0, test_utils_1.fetchMethod)()).toBe('PUT');
            expect(subject.enabled).toBe(false);
        }));
    });
    describe('addFlagTarget', () => {
        it('adds a target to a feature flag', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)({}, { status: 204 });
            yield workos.featureFlags.addFlagTarget({
                slug: 'advanced-dashboard',
                targetId: 'user_01EHQMYV6MBK39QC5PZXHY59C5',
            });
            expect((0, test_utils_1.fetchURL)()).toContain('/feature-flags/advanced-dashboard/targets/user_01EHQMYV6MBK39QC5PZXHY59C5');
            expect((0, test_utils_1.fetchMethod)()).toBe('POST');
        }));
        it('adds an organization target to a feature flag', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)({}, { status: 204 });
            yield workos.featureFlags.addFlagTarget({
                slug: 'advanced-dashboard',
                targetId: 'org_01EHQMYV6MBK39QC5PZXHY59C5',
            });
            expect((0, test_utils_1.fetchURL)()).toContain('/feature-flags/advanced-dashboard/targets/org_01EHQMYV6MBK39QC5PZXHY59C5');
            expect((0, test_utils_1.fetchMethod)()).toBe('POST');
        }));
    });
    describe('removeFlagTarget', () => {
        it('removes a target from a feature flag', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)({}, { status: 204 });
            yield workos.featureFlags.removeFlagTarget({
                slug: 'advanced-dashboard',
                targetId: 'user_01EHQMYV6MBK39QC5PZXHY59C5',
            });
            expect((0, test_utils_1.fetchURL)()).toContain('/feature-flags/advanced-dashboard/targets/user_01EHQMYV6MBK39QC5PZXHY59C5');
            expect((0, test_utils_1.fetchMethod)()).toBe('DELETE');
        }));
        it('removes an organization target from a feature flag', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)({}, { status: 204 });
            yield workos.featureFlags.removeFlagTarget({
                slug: 'advanced-dashboard',
                targetId: 'org_01EHQMYV6MBK39QC5PZXHY59C5',
            });
            expect((0, test_utils_1.fetchURL)()).toContain('/feature-flags/advanced-dashboard/targets/org_01EHQMYV6MBK39QC5PZXHY59C5');
            expect((0, test_utils_1.fetchMethod)()).toBe('DELETE');
        }));
    });
});
