// @workos-inc/node/lib/feature-flags/feature-flags.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeatureFlags = void 0;
const pagination_1 = require("../common/utils/pagination");
const serializers_1 = require("./serializers");
const fetch_and_deserialize_1 = require("../common/utils/fetch-and-deserialize");
class FeatureFlags {
    constructor(workos) {
        this.workos = workos;
    }
    listFeatureFlags(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return new pagination_1.AutoPaginatable(yield (0, fetch_and_deserialize_1.fetchAndDeserialize)(this.workos, '/feature-flags', serializers_1.deserializeFeatureFlag, options), (params) => (0, fetch_and_deserialize_1.fetchAndDeserialize)(this.workos, '/feature-flags', serializers_1.deserializeFeatureFlag, params), options);
        });
    }
    getFeatureFlag(slug) {
        return __awaiter(this, void 0, void 0, function* () {
            const { data } = yield this.workos.get(`/feature-flags/${slug}`);
            return (0, serializers_1.deserializeFeatureFlag)(data);
        });
    }
    enableFeatureFlag(slug) {
        return __awaiter(this, void 0, void 0, function* () {
            const { data } = yield this.workos.put(`/feature-flags/${slug}/enable`, {});
            return (0, serializers_1.deserializeFeatureFlag)(data);
        });
    }
    disableFeatureFlag(slug) {
        return __awaiter(this, void 0, void 0, function* () {
            const { data } = yield this.workos.put(`/feature-flags/${slug}/disable`, {});
            return (0, serializers_1.deserializeFeatureFlag)(data);
        });
    }
    addFlagTarget(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const { slug, targetId } = options;
            yield this.workos.post(`/feature-flags/${slug}/targets/${targetId}`, {});
        });
    }
    removeFlagTarget(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const { slug, targetId } = options;
            yield this.workos.delete(`/feature-flags/${slug}/targets/${targetId}`);
        });
    }
}
exports.FeatureFlags = FeatureFlags;
